ggml/src/ggml-vulkan/vulkan-shaders/mul_mm_embed.comp

```
@@ -0,0 +1,167 @@
+#version 450
+
+#extension GL_EXT_shader_16bit_storage : require
+#extension GL_EXT_shader_explicit_arithmetic_types_float16 : require
+#extension GL_EXT_control_flow_attributes : require
+
+#ifdef A_TYPE_FP16
+    #define A_VEC4_TYPE f16vec4
+    #define A_SCALAR_TYPE float16_t
+    #define A_VEC4_ZERO f16vec4(0.0hf)
+    #define A_VEC4_CAST(v) vec4(v)
+#else
+    #define A_VEC4_TYPE vec4
+    #define A_SCALAR_TYPE float
+    #define A_VEC4_ZERO vec4(0.0f)
+    #define A_VEC4_CAST(v) (v)
+#endif
+
+layout(constant_id = 0) const uint BM = 64;
+layout(constant_id = 1) const uint BN = 64;
+layout(constant_id = 2) const uint BK = 16;
+layout(constant_id = 3) const uint TM = 4;
+layout(constant_id = 4) const uint TN = 8;
+
+const uint WG_X = BN / TN;
+const uint WG_Y = BM / TM;
+const uint WG_SIZE = WG_X * WG_Y;
+const uint VEC_K = BK / 4;
+
+layout(local_size_x = 16, local_size_y = 8, local_size_z = 1) in;
+
+layout (binding = 0) readonly buffer A_BUFFER { A_SCALAR_TYPE data_a[]; };
+layout (binding = 1) readonly buffer B_BUFFER { float         data_b[]; };
+layout (binding = 2) writeonly buffer D_BUFFER { float         data_d[]; };
+
+layout (push_constant) uniform parameter
+{
+    uint M;
+    uint N;
+    uint K;
+    uint stride_a;
+    uint stride_b;
+    uint stride_d;
+} p;
+
+shared A_VEC4_TYPE buf_a[BM][VEC_K];
+shared vec4        buf_b[BN][VEC_K];
+
+void main() {
+    const uint lidx = gl_LocalInvocationID.x;
+    const uint lidy = gl_LocalInvocationID.y;
+    const uint lid = lidy * WG_X + lidx;
+
+    const uint group_m = gl_WorkGroupID.x * BM;
+    const uint group_n = gl_WorkGroupID.y * BN;
+
+    float sums[TM][TN];
+    [[unroll]]
+    for (uint i = 0; i < TM; i++) {
+        [[unroll]]
+        for (uint j = 0; j < TN; j++) {
+            sums[i][j] = 0.0f;
+        }
+    }
+
+    const uint num_k_tiles = (p.K + BK - 1) / BK;
```

I think this is not robust enough and might be wrong for adreno case, but it passes the tests on `test-backend-ops` feel like it shouldn't

---

Shouldn't be hard to add a case or two with odd K. I suggest having relatively small M,N to avoid the error being hidden.